### PR TITLE
Ruby: Add additional sanitizers for Kernel.open or IO.read or similar sinks with a non-constant value

### PR DESCRIPTION
**Pull request by p-**
This PR adds additional sanitizers for the `rb/non-constant-kernel-open` query:
* cases where `freeze` is used on a constant before it's passed to the kernel.open sinks.
* instances where a Kernel method (namely `open`) is overriden and called without arguments.

/cc @erik-krogh

### CODE REVIEW

File: ruby/ql/src/queries/security/cwe-078/NonConstantKernelOpen.ql [hunk 1]
```diff
@@ -18,23 +18,33 @@
...
   or
   // not a concatenation that starts with a constant string
-  hasConstantPrefix(e.(AddExpr).getLeftOperand())
+  exists(DataFlow::ExprNode prefix |
+    node.asExpr().getExpr().(AddExpr).getLeftOperand() = prefix.asExpr().getExpr() and
+    hasConstantPrefix(prefix)
+  )
+  or
+  // is a .freeze call on a constant string
+  node.asExpr().getExpr().(ConstantReadAccess).getValue().(MethodCall).getMethodName() = "freeze"
```
Comment (erik-krogh) [inline ruby/ql/src/queries/security/cwe-078/NonConstantKernelOpen.ql:49]: Shouldn't this be a recursive call to `hasConstantPrefix`.  
Something like: 
```suggestion
  // is a .freeze call on a constant string
  exists(DataFlow::CallNode call | call.getMethodName() = "freeze" | hasConstantPrefix(call.getReceiver()))
```

Comment (erik-krogh) [inline ruby/ql/src/queries/security/cwe-078/NonConstantKernelOpen.ql:49]:
```suggestion
  exists(DataFlow::CallNode call | node = call and call.getMethodName() = "freeze" |
    hasConstantPrefix(call.getReceiver())
  )
  or
  // is a constant read of a constant string
  exists(DataFlow::Node constant |
    constant.asExpr().getExpr() = node.asExpr().getExpr().(ConstantReadAccess).getValue() and
    hasConstantPrefix(constant)
  )
```

I figured out why my approach didn't work.  
1: I forgot to mention `node` in my case for `freeze` :doh: 
2: We need a recursive case for `ConstantReadAccess`, which enables us to find the definition of `CONSTANT_WITH_FREEZE` from where it's used.

File: ruby/ql/src/queries/security/cwe-078/NonConstantKernelOpen.ql [hunk 2]
```diff
@@ -18,23 +18,33 @@
...
+predicate hasConstantPrefix(DataFlow::Node node) {
+  hasConstantPrefix(node.getALocalSource())
+  or
   // if it's a format string, then the first argument is not a constant string
-  e.(StringlikeLiteral).getComponent(0) instanceof StringTextComponent
+  node.asExpr().getExpr().(StringlikeLiteral).getComponent(0) instanceof StringTextComponent
   or
   // it is not a constant string argument
-  exists(e.getConstantValue())
+  exists(node.asExpr().getExpr().getConstantValue())
```
Comment (erik-krogh) [inline ruby/ql/src/queries/security/cwe-078/NonConstantKernelOpen.ql:40]:
```suggestion
  exists(node.getConstantValue())
```

**Review by erik-krogh — APPROVED** [review]
LGTM 👍 

[Evaluation looks good](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/PR-12856-0-ruby__1/reports).  

Can you merge this or do you need me to press the button?

**Comment by p-** [comment]
Thanks!

> Can you merge this or do you need me to press the button?

I cannot press the button, could you do it for me? Thanks

**Pull request merged**